File: services/academic-assessment-service/src/test/java/com/visor/school/assessment/repository/AssessmentRepositoryTest.java
package com.visor.school.assessment.repository;

import com.visor.school.assessment.model.Assessment;
import com.visor.school.assessment.model.AssessmentStatus;
import com.visor.school.assessment.model.AssessmentType;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.test.context.ActiveProfiles;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

@DataJpaTest
@ActiveProfiles("test")
class AssessmentRepositoryTest {

    @Autowired
    private AssessmentRepository assessmentRepository;

    @Test
    void shouldSaveAndFindAssessmentById() {
        // Given
        Assessment assessment = new Assessment(
            UUID.randomUUID(),
            "Test Assessment",
            AssessmentType.TEST,
            new BigDecimal("100.0"),
            UUID.randomUUID()
        );

        // When
        Assessment saved = assessmentRepository.save(assessment);
        Optional<Assessment> found = assessmentRepository.findById(saved.getId());

        // Then
        assertTrue(found.isPresent());
        assertEquals("Test Assessment", found.get().getName());
    }

    @Test
    void shouldFindAssessmentsByClassId() {
        // Given
        UUID classId = UUID.randomUUID();
        Assessment assessment1 = new Assessment(
            classId,
            "Test 1",
            AssessmentType.TEST,
            new BigDecimal("100.0"),
            UUID.randomUUID()
        );
        Assessment assessment2 = new Assessment(
            classId,
            "Test 2",
            AssessmentType.QUIZ,
            new BigDecimal("50.0"),
            UUID.randomUUID()
        );

        assessmentRepository.save(assessment1);
        assessmentRepository.save(assessment2);

        // When
        List<Assessment> found = assessmentRepository.findByClassId(classId);

        // Then
        assertEquals(2, found.size());
    }

    @Test
    void shouldFindAssessmentsByStatus() {
        // Given
        Assessment assessment1 = new Assessment(
            UUID.randomUUID(),
            "Draft Test",
            AssessmentType.TEST,
            new BigDecimal("100.0"),
            UUID.randomUUID()
        );
        Assessment assessment2 = new Assessment(
            UUID.randomUUID(),
            "Published Test",
            AssessmentType.TEST,
            new BigDecimal("100.0"),
            UUID.randomUUID()
        );
        assessment2.publish();

        assessmentRepository.save(assessment1);
        assessmentRepository.save(assessment2);

        // When
        List<Assessment> published = assessmentRepository.findByStatus(AssessmentStatus.PUBLISHED);

        // Then
        assertTrue(published.stream().anyMatch(a -> a.getName().equals("Published Test")));
        assertFalse(published.stream().anyMatch(a -> a.getName().equals("Draft Test")));
    }
}

File: services/academic-assessment-service/src/test/kotlin/com/visor/school/assessment/repository/AssessmentRepositoryTest.kt
package com.visor.school.assessment.repository

import com.visor.school.assessment.model.Assessment
import com.visor.school.assessment.model.AssessmentStatus
import com.visor.school.assessment.model.AssessmentType
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.test.context.ActiveProfiles
import java.math.BigDecimal
import java.util.UUID

@DataJpaTest
@ActiveProfiles("test")
class AssessmentRepositoryTest @Autowired constructor(
    private val assessmentRepository: AssessmentRepository
) {

    @Test
    fun `should save and find assessment by id`() {
        // Given
        val assessment = Assessment(
            classId = UUID.randomUUID(),
            name = "Test Assessment",
            type = AssessmentType.TEST,
            totalPoints = BigDecimal("100.0"),
            createdBy = UUID.randomUUID()
        )

        // When
        val saved = assessmentRepository.save(assessment)
        val found = assessmentRepository.findById(saved.id)

        // Then
        assertTrue(found.isPresent)
        assertEquals("Test Assessment", found.get().name)
    }

    @Test
    fun `should find assessments by class id`() {
        // Given
        val classId = UUID.randomUUID()
        val assessment1 = Assessment(
            classId = classId,
            name = "Test 1",
            type = AssessmentType.TEST,
            totalPoints = BigDecimal("100.0"),
            createdBy = UUID.randomUUID()
        )
        val assessment2 = Assessment(
            classId = classId,
            name = "Test 2",
            type = AssessmentType.QUIZ,
            totalPoints = BigDecimal("50.0"),
            createdBy = UUID.randomUUID()
        )

        assessmentRepository.save(assessment1)
        assessmentRepository.save(assessment2)

        // When
        val found = assessmentRepository.findByClassId(classId)

        // Then
        assertEquals(2, found.size)
    }

    @Test
    fun `should find assessments by status`() {
        // Given
        val assessment1 = Assessment(
            classId = UUID.randomUUID(),
            name = "Draft Test",
            type = AssessmentType.TEST,
            totalPoints = BigDecimal("100.0"),
            createdBy = UUID.randomUUID()
        )
        val assessment2 = Assessment(
            classId = UUID.randomUUID(),
            name = "Published Test",
            type = AssessmentType.TEST,
            totalPoints = BigDecimal("100.0"),
            createdBy = UUID.randomUUID()
        )
        assessment2.publish()

        assessmentRepository.save(assessment1)
        assessmentRepository.save(assessment2)

        // When
        val published = assessmentRepository.findByStatus(AssessmentStatus.PUBLISHED)

        // Then
        assertTrue(published.any { it.name == "Published Test" })
        assertFalse(published.any { it.name == "Draft Test" })
    }
}

